### PR TITLE
Fixed #20705 -- Made PasswordResetForm usable for user models that use an email field not named 'email'.

### PR DESCRIPTION
**Pull request by mlevental**
https://code.djangoproject.com/ticket/20705

### CODE REVIEW

**Comment by timgraham** [comment]
Did you look at my draft patch in https://code.djangoproject.com/attachment/ticket/20705/20705-initial.diff? Despite what I initially wrote on the ticket and even in that patch, it feels a bit inappropriate to have to specify this as part of the form rather than a design which retrieves the email field from the user model (which could presumably be reused in other places).

For example, how about adding `AbstractUser.EMAIL_FIELD` (similar to `USERNAME_FIELD`)? Then the field could be retrieved with something like: 
```
UserModel = get_user_model()
try:
    email_field_name = UserModel.EMAIL_FIELD
except AttributeError:
   email_field_name = 'email'
```
Maybe the try/except could be factored into a `get_email_field_name()` method on `AbstractBaseUser`.

**Comment by timgraham** [comment]
Documentation for `EMAIL_FIELD` is needed. Did you check if there's anywhere else that can use that option?

**Comment by mlevental** [comment]
Yes and I don't think there is another place that option can be used.

**Comment by timgraham** [comment]
buildbot, test this please.

**Comment by timgraham** [comment]
Take a look at the [https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/submitting-patches/#patch-review-checklist patch review checklist]. In particular, I see release notes and a `.. versionadded: 1.11` annotation are missing.

File: docs/releases/1.11.txt [hunk 1]
```diff
@@ -254,6 +254,9 @@ Forms
   <django.forms.Form.get_initial_for_field>` method returns initial data for a
   form field.
 
+* The ``PasswordResetForm.get_users()`` method now supports a custom email
```
Comment (timgraham) [inline docs/releases/1.11.txt]: PasswordResetForm supports custom user models that use an email field named something other than 'email'. Set `CustomUser.EMAIL_FIELD` to the name of the field.
(add links as appropriate)

File: docs/topics/auth/customizing.txt [hunk 2]
```diff
@@ -547,6 +547,12 @@ password resets. You must then provide some key implementation details:
         value (the :attr:`~django.db.models.Field.primary_key` by default) of an
         existing instance.
 
+    .. attribute:: EMAIL_FIELD
+
+        .. versionadded:: 1.11
+
+        A string describing the name of the email field on the ``User`` model.
```
Comment (timgraham) [inline docs/topics/auth/customizing.txt]: Probably want to add a link to `get_email_field_name()` here once that method is documented.

File: docs/topics/auth/customizing.txt [hunk 3]
```diff
@@ -810,9 +816,10 @@ The following forms make assumptions about the user model and can be used as-is
 if those assumptions are met:
 
 * :class:`~django.contrib.auth.forms.PasswordResetForm`: Assumes that the user
-  model has a field named ``email`` that can be used to identify the user and a
-  boolean field named ``is_active`` to prevent password resets for inactive
-  users.
+  model has a field that can be used to identify the user. By default the
```
Comment (timgraham) [inline docs/topics/auth/customizing.txt]: default,

File: docs/topics/auth/customizing.txt [hunk 4]
```diff
@@ -810,9 +816,10 @@ The following forms make assumptions about the user model and can be used as-is
...
+  expected field name is ``email`` which can be changed by setting the ``User``
+  model field ``EMAIL_FIELD``. Additionally a boolean field named ``is_active``
```
Comment (timgraham) [inline docs/topics/auth/customizing.txt]: model field -> model attribute? (add a link?)
Additionally,

File: tests/auth_tests/models/with_custom_email_field.py [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+        return user
+
+    def get_by_natural_key(self, username):
+        return self.get(username=username)
+
+
+class CustomEmailField(AbstractBaseUser):
+    username = models.CharField(max_length=255)
+    password = models.CharField(max_length=255)
+    email_field_name = models.EmailField()
```
Comment (timgraham) [inline tests/auth_tests/models/with_custom_email_field.py]: email_field_name -> email_address to make it more realistic

**Comment by timgraham** [comment]
This patch is small enough, that you can squash commits as you go and force push to your branch.

**Comment by timgraham** [comment]
merged in 617e36dc1ede2a311dfd03f18432b31cbfe4c0f7, thanks!

**Pull request closed without merging**